src/vos/vos_query.c

```diff
@@ -203,25 +198,53 @@ query_recx(struct open_query *query, daos_recx_t *recx)
 	if (rc != 0)
 		goto fini;
 
-	recx->rx_idx = entry.en_sel_ext.ex_lo;
-	recx->rx_nr = entry.en_sel_ext.ex_hi - entry.en_sel_ext.ex_lo + 1;
+	if (opc & EVT_ITER_SKIP_HOLES) {
+		if (entry.en_visibility & EVT_VISIBLE) {
+			recx->rx_idx = entry.en_sel_ext.ex_lo;
+			recx->rx_nr = entry.en_sel_ext.ex_hi - entry.en_sel_ext.ex_lo + 1;
+		} else {
+			rc = -DER_NONEXIST;
+		}
+		D_GOTO(fini, rc);
+	}
+
+	/* The following only supports MAX recx query at the moment */
+	if (!(query->qt_flags & VOS_GET_MAX))
+		D_GOTO(fini, rc = -DER_INVAL);
+
+	/* Check if these extents can be merged */
+	D_ASSERT(bio_addr_is_hole(&entry.en_addr));
```

Comment:
when no EVT_ITER_SKIP_HOLES falg, not sure if can assert here, if there are with both valid ext and hole ext.
may confirm with Jeff.

Reply:
I think it's safe because the only way you get here happens to be when EVT_ITER_SKIP_DATA is also set.   Maybe add a comment here or an assertion to that effect so if someone ever breaks this prerequisite, it's more obvious.